Automattic/woocommerce-payments
# Add JSON rule builder for the rule engine

Fixes Automattic/woocommerce-payments-server#3043

#### Changes proposed in this Pull Request

This adds the communication layer between the fraud protection settings and the server by creating a ruleset on the UI itself and sending it to the server as is, and also handles the storage of the local copy of it. The local copy will last 1 day, and will updated when the cache expires (uses WP transients). 

Also, this PR contains the part which saves the fraud protection level, and modifies the ruleset on the server according to the setting selected. 

This PR also includes: 
- the inclusion of the Basic protection level as the default, 
- the indication of AVS and CVC checks according to Stripe account settings on the advanced protection settings cards,
- the escalation of protection level to advanced, when the settings are saved from the advanced protection settings page.
- the removal of order velocity rule, and the fixes for the fraud protection levels help modals according to p1679314937335329-slack-C01B8KNUYSW.









#### Testing instructions




**Preparation**
* Checkout this branch
* Run `npm run build:client` to update the client scripts
* Open your client DB and check if there are 
    * `_transient_wcpay_fraud_protection_settings` 
    * `_transient_timeout_wcpay_fraud_protection_settings`
    *  `current_protection_level`
options in `wp_options` table. If they exists, delete them. Also check if you have the feature flag (`wcpay_fraud_protection_settings_active`) enabled (set as `1`) to see the UI on the same table. 

    You can use `SELECT * FROM wp_options where option_name in ('current_protection_level', '_transient_wcpay_fraud_protection_settings', '_transient_timeout_wcpay_fraud_protection_settings');`
to query the necessary things to delete.
* Open your server DB and empty the `wcpay_fraud_rulesets` table.

**Testing**

*Initializing the ruleset config on the server*

* When you open a page that requests the settings, 
    * the protection level option should be set as **basic**, 
    * the transient will contain an empty array, 
    * and the same ruleset (empty array) will be stored on the server for the account. 
    * And if you refresh that page, no duplicate records should be created, the system will use the transient value.

*Updating the settings with presets*

* When you change the fraud protection level on the _Payments > Settings_ page, 
    * the **current_protection_level** should change accordingly.
    * the transient will contain the relevant settings, matching with the settings shown in the relevant modal.
    * the server ruleset table will contain a new row, having a matching ruleset with the transient value.
    * When you select Advanced and click Save Settings, the last ruleset definition should stay intact, and only the advanced level will be set on the related option.
    * When you click Configure, you should see the last selected preset's values on the form.

*Updating the settings with advanced definitions*
* While you have a preset level (something different than the Advanced level) saved as your protection level, click the **Advanced** radio button and click **Configure**. Modify your settings a bit, and click **Save changes**.
* You should see a **Current protection level is set to "advanced".** message just before the **Settings Saved** message.
    * the protection level option should be set as **advanced**, 
    * the transient will contain the matching ruleset, 
    * and the same ruleset will be stored on the server for the account. 
    * And if you click "Save changes" again on that page without changing anything, no duplicate records should be created on the server.

_Testing synchronization errors_
1. The client loses the configuration
* Delete the options and transients from the client DB. 
* Refresh the advanced settings page, and see if the settings are fetched back from the server.
2. The server loses the configuration
* Delete all the rulesets for the current account on the DB
* Set the `_transient_timeout_wcpay_fraud_protection_settings` to `0`, to simulate transient expiration
* Refresh the advanced settings page
* Check the server DB to see if the client ruleset settings are saved back to the server ruleset table.
3. The client and server has different settings
* Select and save some presets from the Payments > Settings page, then select Advanced, save, and go to the advanced settings page. Modify some settings on the form and save it.
* Delete the last saved ruleset for the current account on the DB.
* Set the `_transient_timeout_wcpay_fraud_protection_settings` to `0`, to simulate transient expiration
* Refresh the advanced settings page
* Check if the advanced settings page now contains the last ruleset record on the server DB fraud rulesets table. (The source of truth needs to be the server on this case)

-------------------


**Post merge**
